### PR TITLE
Webinterface 2.0

### LINKED ISSUE #352: Websockets
As a player, I want the webserver to make use of sockets, so I have a quicker response when I perform an action.

This task has the following subtasks:
- #353 Websockets - Finalize implementation - Server
- #354 Websockets - Finalize implementation - Client
- #355 Websockets - Test websockets

### LINKED ISSUE #353: Websockets - Finalize implementation - Server
Finalize the java side of the socket implementation.

This is a subtask of #352 - Websockets

### LINKED ISSUE #354: Websockets - Finalize implementation - Client
Finalize the javascript side of the socket implementation.

This is a subtask of #352 - Websockets

### LINKED ISSUE #355: Websockets - Test websockets
After all code has been finalized the websockets need to be tested.

This is a subtask of #352 - Websockets

### LINKED ISSUE #359: Make webinterface user friendly - Alerts
Show alerts when an action fails.

This is a subtask of #356 - Make webinterface user friendly

### LINKED ISSUE #408: Rewrite of web interface
In addition to the addition of websockets (see #352), we want to completely rewrite the web interface, as we are not really satisfied with it in it's current state.

The new web interface must be more robust. That means:
- Clients must be able to close and reopen the page without issue.
- The server must be able to be restarted without requiring a page reload
- If the browser does not support websockets, the web interface will use a different method.

### PR DESCRIPTION
**Pull request by Taeir**
Completely overhauled the webinterface.

It might not look that pretty, but it's very solid. It doesn't break (I encourage you to try it).
- If you close the page and open it again, you will be put back in the game.
- If the websocket connection is closed by a server error, you will be
  INSTANTLY reconnected.
- If the server is stopped, you will be sent to the menu, where the page will keep on trying to reconnect. The join game button will be disabled.
- If the server is started again, the join button will be enabled again.
- If you have a bad browser that doesn't support websockets, everything still works

Other changes:
- [BOTH] Added websockets
- [CLIENT] Split all the javascript code into multiple .js files.
- [BOTH] Cleaned up and refactored all code.
- [CLIENT] Changed for(i = =;...) to for(VAR i = 0;...) to prevent concurrency issues.
- [CLIENT] Added logo to index page
- [BOTH] Fixed switching from dwarfs to elves not clearing exploration.
- [SERVER] Added a lot more tests.
- [CLIENT] We are now using alerts (#359).

Coverage

This closes #352, closes #353, closes #354, closes #355, closes #408, closes #359.

_"Sturdy like a brick, elegant like a brick"_

### CODE REVIEW

**Comment by Geweldig** [comment]
Looks great, should be merged.

File: src/main/java/nl/tudelft/contextproject/webinterface/Team.java [hunk 1]
```diff
@@ -0,0 +1,10 @@
+package nl.tudelft.contextproject.webinterface;
+
+/**
+ * Enum for the different teams.
+ */
+public enum Team {
+	DWARFS,
+	ELVES,
+	NONE;
+}
```
Comment (Diocruel) [inline src/main/java/nl/tudelft/contextproject/webinterface/Team.java:10]: Jurgen will finally be happy now :+1:

File: src/main/resources/webinterface/js/handlers.js [hunk 2]
```diff
@@ -0,0 +1,278 @@
...
+            return;
+        }
+    } else if (evt.data.length === 3) {
+        if (ErrorCodes[evt.data] !== undefined) {
+            handleErrorCode(evt.data);
+            return;
+        }
+    }
+    
+    //TODO try catch?
```
Comment (Diocruel) [inline src/main/resources/webinterface/js/handlers.js:164]: This is not really descriptive what TODO here

File: src/main/resources/webinterface/js/nosockets.js [hunk 3]
```diff
@@ -0,0 +1,513 @@
...
+function switchTo(view) {
+    //We don't need to switch to our current state.
+    if (gView == view) return;
+    
+    console.log("[DEBUG] Switching to " + view);
+    gView = view;
+    requestMap();
+    switch (view) {
+        case "WAITING":
+            //TODO
```
Comment (Diocruel) [inline src/main/resources/webinterface/js/nosockets.js]: TODO what here?

File: src/main/resources/webinterface/js/nosockets.js [hunk 4]
```diff
@@ -0,0 +1,513 @@
...
+            break;
+        case "RUNNING":
+            $(document.getElementById("selectTeam")).hide();
+            break;
+        case "PAUSED":
+            console.log("PAUSED STATE HIT");
+            break;
+        case "ENDED":
+            //TODO
```
Comment (Diocruel) [inline src/main/resources/webinterface/js/nosockets.js]: TODO what here?

File: src/main/resources/webinterface/js/requests.js [hunk 5]
```diff
@@ -0,0 +1,280 @@
...
+/**
+ * Requests the server to set the team, either via websocket or normally.
+ *
+ * @param team
+ *      the team to set
+ */
+function requestSetTeam(team) {
+    requestNormal_setTeam(team);
+    
+    //TODO WebSocket
```
Comment (Diocruel) [inline src/main/resources/webinterface/js/requests.js:115]: Aren't they done in this PR? xD

**Comment by Diocruel** [comment]
The rests of the code looks great! Mergable after my comments have been addressed.

File: src/main/resources/webinterface/js/handlers.js (hunk 2 above)
Comment (Taeir) [inline src/main/resources/webinterface/js/handlers.js:164]: TODO: Add try catch if playtesting shows that problems occur

**Comment by Diocruel** [comment]
Okay can merge after Travis. My comments have been addressed in real life discussion.

File: src/main/resources/webinterface/js/requests.js (hunk 5 above)
Comment (Taeir) [inline src/main/resources/webinterface/js/requests.js:115]: The websocket implementation of setTeam is not done. We might not need it. Testing must show.

**Pull request merged**